src/main/java/org/broadinstitute/hellbender/tools/exome/ExonDB.java

```diff
@@ -0,0 +1,299 @@
+package org.broadinstitute.hellbender.tools.exome;
+
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.GenomeLoc;
+import org.broadinstitute.hellbender.utils.IndexRange;
+
+import java.util.List;
+
+/**
+ * Exon collection.
+ * <p>
+ *      This interface includes operations to query exons based on genomic coordinates and their names (if provided).
+ * </p>
+ *
+ * <p>
+ *      Additional per exon meta-data can be added by implementing class through the customizable type parameter
+ *      &lt;E&gt;.
+ * </p>
+ *
+ * @param <E> exon meta-data type.
+ * @author Valentin Ruano-Rubio &lt;valentin@broadinstitute.org&gt;
+ */
+public interface ExonDB<E> {
+
+    /**
+     * Returns the number of exons in this Exome.
+     *
+     * @return 0 or greater.
+     */
+    int exonCount();
+
+    /**
+     * Returns a exon given is index in the exon.
+     * <p>
+     * Indices are 0-based, thus the first exon has index 0, the second 1 and so forth.
+     * </p>
+     *
+     * @param index the query index.
+     * @throws IndexOutOfBoundsException if {@code index} is not within
+     *                                   valid bounds <code>[0 .. ({@link #exonCount()})</code>.
+     */
+    E exon(final int index);
+
+    /**
+     * Returns the exon given its name.
+     *
+     * @param name the query exon name.
+     * @return {@code null} if there is not such an exon.
+     */
+    default E exon(final String name) {
+        final int index = index(name);
+        if (index < 0) {
+            return null;
+        } else {
+            return exon(index);
+        }
+    }
+
+    /**
+     * Returns the exon index given its name.
+     *
+     * @param name the query exon name.
+     * @throws IllegalArgumentException if {@code name} is {@code null}.
+     * @return {@code -1} if there is not such an exon.
+     */
+    int index(final String name);
+
+    /**
+     * Returns the exon that overlap a particular location.
+     * <p>
+     *     The returned exon coordinates does not need to match exactly the
+     * the query {@code location} just overlap it by at least one base</p>.
+     * <p>
+     *     This method will return {@code null} if there is no such a exon.
+     * </p>
+     * <p>
+     *     If there is more than on exon that overlap with the given region it will
+     *     throw a {@link AmbiguousExonException} instead.
+     * </p>
+     *
+     * @param overlapRegion the query location.
+     * @return {@code null} if there is no overlapping exons at {@code overlapRegion}.
+     * @throws IllegalArgumentException if {@code overlapRegion} is {@code null}.
+     * @throws AmbiguousExonException if the query location overlaps more than one exon.
+     */
+    default E exon(final GenomeLoc overlapRegion) {
```

Comment:
I suggest moving away from GenomeLoc if only possible - we don't need a reference here and GenomeLoc is on the kill list. @droazen and @lbergelson can help on the alternatives.